compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -667,29 +667,36 @@ trait Implicits { self: Typer =>
         EmptyTree
     }
 
-    def synthesizedTypeTag(formal: Type): Tree = formal.argInfos match {
-      case arg :: Nil if !arg.typeSymbol.is(Param) =>
-        object bindFreeVars extends TypeMap {
-          var ok = true
-          def apply(t: Type) = t match {
-            case t @ TypeRef(NoPrefix, _) =>
-              inferImplicit(defn.QuotedTypeType.appliedTo(t), EmptyTree, span) match {
-                case SearchSuccess(tag, _, _) if tag.tpe.isStable =>
-                  tag.tpe.select(defn.QuotedType_splice)
-                case _ =>
-                  ok = false
-                  t
-              }
-            case _ => t
+    def synthesizedTypeTag(formal: Type): Tree = {
+      def quotedType(t: Type) = {
+        if (StagingContext.level == 0)
+          ctx.compilationUnit.needsStaging = true // We will need to run ReifyQuotes
+        ref(defn.InternalQuoted_typeQuote).appliedToType(t)
```

Comment:
I'm not sure if _implicit search_ is the best place to put the logic --- what about put the logic in typer?

Reply:
We detect quotes and splices when we type. The issue here is that we will do not type this tree because we create it already typed.

Reply:
The concern I have here is about _modularity_. It seems the added code does not belong here. Is it due to concern about performance? Modularity could be sacrificed for performance.